### DISCUSSION TITLE
Should batch_run default to running on 1 CPU?

### DISCUSSION THREAD

**Discussion by rht** [mesa/mesa#1202] [General]
Defaulting to 1 CPU will simplify the tutorial on Windows.

Then, if `number_processes` is > 1 and on Windows, we show a warning that tells user to add `freeze_support()` (we pre-emptively tell the user of this issue instead of waiting for it to happen). This warning can be disabled via a flag to the function:
```python
batch_run(
    ...
   hide_windows_freeze_support_warning=True,
   ...
)
```

**Comment by tpike3** [comment]
I think this is the way to go...particularly with new users this is adding cognitive overload. Better to simplify it.

**Comment by tpike3** [comment]
Looking at different `setup.py` for #1215, I noticed this use of [OpenMP](https://github.com/scikit-learn/scikit-learn/blob/2c5058e0d6cc4f5c627d2ee256f4735fd5ba4a39/setup.py#L123) in SciKit Learn. I will explore more as I wonder if there may be solution in this approach to our multiprocessing issues.

**Reply by rht** [reply]
They seem to use it for the purpose of CI. I have never observed the use of shell env vars in Python scientific computing. The closest one is setting the port for a Flask server, but this is not scientific computing.

**Comment by clarkpetri** [comment]
I'm running on a Mac M1 and also had to set number_processes = 1 to prevent throwing the errors described in the tutorial. I originally thought I wouldn't need to do so due to working in MacOS.

**Reply by rht** [reply]
Huh, that's good to know. Accordiing to https://docs.python.org/3/library/multiprocessing.html#multiprocessing.freeze_support

> Calling freeze_support() has no effect when invoked on any operating system other than Windows.

I have read an issue thread somewhere in https://github.com/scipy/scipy that on macOS, it's not the `freeze_support()` that fixes the problem, it is actually due to simply importing multiprocessing at the parent file. Would be great if you could test simply `import multiprocessing`, without the `freeze_support()`.

**Reply by clarkpetri** [reply]
Added `import multiprocessing` to the top of the code and set `number_processes=None` and had the issue return. I'm happy to test other things to assist the community. My python installation is operating natively on Apple silicon and not through Rosetta if that matters. 

Thanks! I'm loving Mesa and determined to finish my PhD with it as my primary ABM environment.

**Reply by rht** [reply]
OK, I found the relevant comment in the SciPy thread: https://github.com/scipy/scipy/issues/11835#issuecomment-612685404. I also found this [SO question](https://stackoverflow.com/questions/67999589/multiprocessing-with-pool-throws-error-on-m1-macbook) with a possibly working solution, i.e. either by forcing to use the "fork" method, or use the `ray` library.

> Thanks! I'm loving Mesa and determined to finish my PhD with it as my primary ABM environment.

Good luck with your PhD!

**Reply by tpike3** [reply]
@clarkpetri Just to make sure, are you running your model in Jupyter or an IDE (if so which one) or the Python interpreter?

**Reply by tpike3** [reply]
Also go George Mason :)

**Reply by clarkpetri** [reply]
> @clarkpetri Just to make sure, are you running your model in Jupyter or an IDE (if so which one) or the Python interpreter?

Jupyter in this instance for BatchRunner.

It's not immediately clear to me what I would do with the fork method or `ray` library but I'll fiddle with it when I have time (projects, sick kid, very pregnant wife...).

The Mason CSS curriculum has been great to me! Very happy here.

**Reply by tpike3** [reply]
@clarkpetri For batch_run its best not to run in Jupyter.  If you pull the code outside jupyter on a Mac is should run fine, otherwise you need to get the [Jupyter Extension](https://github.com/micahscopes/nbmultitask) so you can run multiprocessing with Jupyter.

**Reply by clarkpetri** [reply]
Good to know. Thank you! Multiprocessing is working now, though I need to figure out how to stop the subsequent lines of code from executing until all multi-thread operations are complete.

**Reply by tpike3** [reply]
Make sure you wrap the whole model in batch_run, it is effectively just putting one model with a specific set of parameters on one core. So yes having code outside the batch_run if it is tied to the model will produce odd behaviors.

**Comment by clarkpetri** [comment]
Got it working now! I don't know what I was doing wrong other than being dense and tired with a sick toddler. Code below and it's performing the 500 runs for my training set substantially faster:
![Screen Shot 2022-04-14 at 10 14 53 AM](https://user-images.githubusercontent.com/99350552/163409606-6c1b98a0-9dd0-4cca-bde8-cee8fb19b7c6.png)
.

**Reply by rht** [reply]
This [SO answer](https://stackoverflow.com/a/70513640) seems to suggest that putting the code inside `if __name__ == "__main__":` is the piece that actually resolves the problem instead of the `freeze_support()`. I wonder if you can test this hypothesis.

**Reply by clarkpetri** [reply]
That's it. Keeping `if __name__ == "__main__":` but commenting out `freeze_support()` worked. The inverse and tabbing the code in threw all the same errors I was experiencing.

I was averaging ~ 7it/s on single core but ~65it/s now. 👍

**Comment by rht** [comment]
Will be resolved by #1210.

**Comment by ymgongzi13** [comment]
if I want to run it in jupyter notebook,  how should I solve this problem?
![image](https://user-images.githubusercontent.com/81600318/226531791-9aa7d451-64cb-4fa1-a643-6f91adabfedb.png)

**Comment by ymgongzi13** [comment]
and if I use the method above, it works like followings:
![image](https://user-images.githubusercontent.com/81600318/226531141-8fecb067-2577-4681-8208-e3dd6ede1062.png)

**Reply by rht** [reply]
You have a typo "mian".

**Reply by ymgongzi13** [reply]
Oh, God! What a stupid error! Maybe my brain needs to take a break. Now it works. Thank you very much!!!!!!!!!!!!!!!!!!!!!!!

**Reply by ymgongzi13** [reply]
Well, I found a problem nowadays.  Though my scripts can run normally, its speed is still very slow when my 'number_processes' was set to 8.
And it need about 2 hours to finish simulate.
However, the same methods in netlogo just spend a few minutes.    
![image](https://user-images.githubusercontent.com/81600318/230642755-e6ac097e-c9fb-420e-b7e2-f76bbc8de79d.png)
Why does this problem happen? And is there some methods to spend less time?
Wait for your kind advice! Thank you very much!

**Reply by tpike3** [reply]
There should not be that big a difference in the Netlogo run and your Mesa run. 

First, are running the same number of iterations? What is your total number of model runs for each version (i.e. Mesa and NetLogo)? 

Second, you can find inefficient code, which with this big a difference would indicate to me a possible bug with performance testing. Here is a [good overview of performance testing](https://granulate.io/blog/python-performance-testing-quick-tutorial-and-best-practices/)

**Reply by ymgongzi13** [reply]
They both have 11000 iterations. 
Maybe my code needs to be simplified.
I'll try to find the inefficient code by the lib you introduced. 
Thanks again for your kind help!!
